src/gc-alloc-profiler.cpp

```diff
@@ -268,18 +87,59 @@ RawBacktrace get_raw_backtrace() {
 // == exported interface ==
 
 JL_DLLEXPORT void jl_start_alloc_profile(int skip_every) {
-    jl_printf(JL_STDERR, "g_alloc_profile se:%d allocs:%d \n", g_alloc_profile.skip_every, g_alloc_profile.allocs.size());
     g_alloc_profile_enabled = true;
-    g_alloc_profile.reset(skip_every);
+    g_alloc_profile = AllocProfile{skip_every};
 }
 
 extern "C" {  // Needed since the function doesn't take any arguments.
 
-JL_DLLEXPORT struct AllocResults jl_stop_and_write_alloc_profile() {
+JL_DLLEXPORT struct RawAllocResults jl_stop_alloc_profile() {
     g_alloc_profile_enabled = false;
 
-    return AllocResults{g_alloc_profile.allocs.size(),
-                    g_alloc_profile.allocs.data()};
+    auto results = RawAllocResults{
+        g_alloc_profile.allocs.data(),
+        g_alloc_profile.allocs.size()
+    };
+    
+    // package up type names
+    results.num_type_names = g_alloc_profile.type_name_by_address.size();
+    // TODO: free this malloc
+    results.type_names = (TypeNamePair*) malloc(sizeof(TypeNamePair) * results.num_type_names);
+    int i = 0;
+    for (auto type_addr_name : g_alloc_profile.type_name_by_address) {
+        jl_printf(JL_STDERR, "type name: %s\n", type_addr_name.second.c_str());
+        // TODO: free this malloc
+        char *name = (char *) malloc(type_addr_name.second.length() + 1);
+        memcpy(name, type_addr_name.second.c_str(), type_addr_name.second.length() + 1);
+        results.type_names[i++] = TypeNamePair{
+            type_addr_name.first,
+            name,
+        };
+    }
```

Comment:
Why do we compute the type names in C++? Can't we do this on the julia side, now? If we just store the type's pointer?

Or is this the thing that was segfaulting when you tried it, and you didn't know why?

Reply:
If we need to go with this approach, one option is we can construct _julia strings_ here, that will be managed by the GC, so they never need to be freed? That way they can be passed directly to the julia side.

Reply:
yeah this is the thing that was segfaulting :/ I can show you, so you can sanity check me… I would very much prefer to just pass back type pointers and not do this.

I initially tried Julia strings, but that didn't work either — I believe they were getting GC'd, because while I could construct them here, I couldn't figure out how to tell the GC to root them.

Reply:
🤔 oh, i see you were trying that in a previous commit.. I think maybe you just weren't using the write string constructor, and you were creating like a _view_ over the char* instead of copying the data out of the char*. I think we should be able to get that to work.....

Reply:
yeah, weird. it really shouldn't segfault. i think that should be fine. I'd be interested to work through it with you! 👍

Reply:
yeah, probably worth another try before we dig deeper into string stuff